src/rt/reshoot.1
.TH RESHOOT 1 BRL-CAD
.\"                          R E S H O O T . 1
.\" BRL-CAD
.\"
.\" Copyright (c) 2006-2012 United States Government as represented by
.\" the U.S. Army Research Laboratory.
.\"
.\" Redistribution and use in source (Docbook format) and 'compiled'
.\" forms (PDF, PostScript, HTML, RTF, etc.), with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\" 1. Redistributions of source code (Docbook format) must retain the
.\" above copyright notice, this list of conditions and the following
.\" disclaimer.
.\"
.\" 2. Redistributions in compiled form (transformed to other DTDs,
.\" converted to PDF, PostScript, HTML, RTF, and other formats) must
.\" reproduce the above copyright notice, this list of conditions and
.\" the following disclaimer in the documentation and/or other
.\" materials provided with the distribution.
.\"
.\" 3. The name of the author may not be used to endorse or promote
.\" products derived from this documentation without specific prior
.\" written permission.
.\"
.\" THIS DOCUMENTATION IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY
.\" EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
.\" OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
.\" BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
.\" LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
.\" USE OF THIS DOCUMENTATION, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\".\".\"
.SH NAME
reshoot \- re-shoot rays and compare results
.SH SYNOPSIS
.B reshoot
model.g objects ... < raylist
.SH DESCRIPTION
This program is designed to shoot a set of rays and compare the results with previous results.
The program exits with a non-zero status if any rays are different from the previous run,
and a 0 status if no differences were found.


.P
Typically, an application (probably other than
.I reshoot\fR)
is run with the RT_G_DEBUG flag DEBUG_ALLHITS set.
This will cause librt to log all calls to the application \fIhit\fR routine to the standard error.
The logged output can be processed by an
.I awk
program to produce input to
.I reshoot\fR.
Consider the file p.awk with the following contents:
.P
.in +5

.nf
/Pnt/ { START=index($0,"(") + 1
       STR=substr($0, START, index($0,")") - START)
       gsub(  ", "  , "," , STR)
       printf "Pnt=%s\\n",STR
       }
/Dir/ { START=index($0,"(") + 1
       STR=substr($0, START, index($0,")") - START)
       gsub(  ", "  , "," , STR)
       printf "Dir=%s\\n",STR
       }
/PT/  { PIN=$3; POUT=$5 }
/InHIT/ { INHIT=$2 }
/OutHIT/ { OUTHIT=$2 }
/Region/ { printf "\tregion=%s in=%s in%s out=%s out%s\\n",
	$2,PIN,INHIT,POUT,OUTHIT
	}
.fi

.P
Given this
.I awk
program The log file from the application can be processed as follows:
.IP
awk -f p.awk < logfile > shotlist
.P
to produce input for
.I reshoot\fR.
An example of the input file format for
.I reshoot
is:
.P
.in +5

.nf
Pnt=0,0,1
Dir=0,0,-1
	region=r1.r in=p1.s indist=10016.8 out=p2.s outdist=10017.8
	region=r2.r in=p3.s indist=10018.8 out=p4.s outdist=10019.8
	\fI... (arbitrary many region lines)\fR
Pnt=0,0,10
Dir=0,0,-1
	region=r1.r in=p1.s indist=10016.8 out=p2.s outdist=10017.8
	region=r2.r in=p3.s indist=10018.8 out=p4.s outdist=10019.8
	\fI... (arbitrary many region lines)\fR
.fi

.P
There can be an arbitrary number of rays specified in the file.
There must not be whitespace between the values of Pnt and Dir.
It is best to comma separate the three coordinate values.
The keywords "Pnt" and "Dir" must occur at the beginning of a line.
There can be an arbitrary number of regions along the ray path.
Each region must be described on a single line.
A new ray is started by a "Pnt" value.
Then
.I reshoot
can be run as follows:
.IP
reshoot geom.g obj ... < shotlist
.P
Differences are reported for each ray as they are found:

.in +5
Ray Pt 0,0,1 Dir 0,0,-1
	inhit mismatch 17.8177 16.8
	outhit mismatch 24.7685 23.8
	region name mismatch /oops.r sb r1.r
	in primitive name mismatch p8.s sb p1.s


.SH BUGS
The floating point values logged when the DEBUG_ALLHITS flag is set in RT_G_DEBUG are printed with a %g format.
This causes some truncation of precision.  This is accounted for by
.I reshoot
by re-formatting the numbers computed into ascii and re-parsing them back in.
This causes the same amount of precision loss as the original values
have undergone.  However, small actual differences are also overlooked
as a result.

.SH AUTHOR
BRL-CAD Team
.SH COPYRIGHT
This software is Copyright (c) 2006-2012 United States Government as
represented by the U.S. Army Research Laboratory. All rights reserved.
.SH "BUG REPORTS"
Reports of bugs or problems should be submitted via electronic
mail to <devs@brlcad.org>.


NEWS
						 -*- coding: utf-8 -*-
BRL-CAD Release Notes
=====================

This document provides a transcript of release notes and user-visible
changes for each release of BRL-CAD.  The individual changes are
listed in chronological order with the most recent changes listed
first and are grouped by release.  Each release also includes optional
descriptive text that emphasizes or further describes significant
changes made.  See document footer for additional details.

----------------------------------------------------------------------
--- 2012-07-XX  Release 7.22.2                                     ---
----------------------------------------------------------------------

* fixed corrupted tessellation pipe export bug - Richard Weiss
* output from g-var no longer prints 'fwrite' messages - Sean Morrison
* fixed crash in step-g tool importing AP214 entities - Sean Morrison
* improved 3dm-g output and invalid geometry support - Sean Morrison
* fixed crash revolving empty sketch objects - Sean Morrison
* fixed TIE raytracing bugs shotlining plate-mode BoTs - Sean Morrison
* g_diff gives non-zero return code if .g files differ - Sean Morrison
* fixed typographical errors in numerous tools and docs - Carl Moore
* fixed TIE perpendicular face mesh evaluation failure - Richard Weiss


----------------------------------------------------------------------
--- 2012-06-25  Release 7.22.0                                     ---
----------------------------------------------------------------------

After many months of development, this release includes hundreds of
fixes for source code defects, logic and memory management bugs,
potential security vulnerabilities, and numerous other issues reported
by Coverity Static Analysis.  While having participated in the
Coverity Scan Initiative since inception in 2006, this release marks a
completion milestone with the review and resolution of all reported
issues, more than 2100 overall.  Every defect was inspected, code
modifications were peer-reviewed, critical features were tested, and
user-visible changes were documented.  Along with other on-going
improvement efforts such as treating all compiler-generated warnings
as errors, time and attention continues to be put forth towards
improving BRL-CAD's maintainability and source code quality.

Development of BREP/NURBS representation support within BRL-CAD
continues to mature with this release including ray tracing
improvements, tighter bounding boxes for better performance, progress
converting from traditional implicit+CSG representations to NURBS+CSG,
wireframe improvements, and a thorough verification and validation
(V&V) review for analysis suitability.  The V&V review compared
BRL-CAD's primitives in implicit, polygonal mesh, and NURBS form
evaluating more than a dozen separate tests looking for differences in
rendering, curvature, presented areas, volume, and more.  The review
identified a couple minor issues that were inspected and fixed but
successfully demonstrated the viability of using our BREP/NURBS
implementation for geometric representation and analysis purposes.

This release introduces numerous major feature developments including
the initial implementation of a new physics simulation capability, a
new ambient occlusion rendering mode, and the integration of rtwizard
into Archer.  The simulation system was developed by Abhijit Nandy
under the European Space Agency's 2011 Summer of Code in Space (SOCIS)
program; and it allows BRL-CAD geometry objects to dynamically
interact with each other (e.g., collision detection, friction) or with
the environment (e.g., gravity).  Ambient occlusion is a shading
method that helps add considerable realism and feature definition to
renderings, enabled within rt using the "set ambSamples=128" command.
The rtwizard rendering interface was extended to work as a
command-line tool (useful for scripting) and was integrated within
Archer as the underlying rendering interface.

Conversion support continues to be a development priority with
significant importer enhancements for the STEP, OBJ, COMGEOM, FASTGEN,
and RAW file formats.  The STEP importer in particular receives
several updates, support for importing assembly hierarchies, and is in
the process of being fully ported to Windows.  This release also fixes
more than 40 user-visible bugs including a significant database
corruption issue.  A database corruption issue was discovered and
fixed affecting 64-bit big endian platforms (e.g., IBM Power Series
supercomputers) when writing out ebm, dsp, vol, and hf primitives.

BRL-CAD's documentation continues to be extended, improved, and
translated due to the sustained efforts of several contributors.
Manual page documentation is now available on Windows through the
'brlman' command and within MGED.  Several utilities and commands that
previously had no documentation now have manual pages; many others
received corrections and improvements.  The introductory article
"About BRL-CAD" is now available in Russian, Italian, and Armenian.

* numerous manual page corrections and improvements - Tom Browder
* integration of rtwizard traits into archer - Bob Parker, Cliff Yapp
* new asc2dsp prep tool for displacement map data files - Tom Browder
* fixed a problem with rt writing pix files on Windows - Bob Parker
* added new raw-g raw triangle mesh to BoT importer - Daniel Roßberg
* added Windows man page support (using html) to brlman - Cliff Yapp
* added support for section numbers to brlman script - Cliff Yapp
* new manual pages for asc2dsp, dsp, asc2pix, and asc2g - Tom Browder
* fixed bug creating binary uniform (binunif) objects - Sean Morrison
* fixed memory corruption bug reading v4 geometry - Sean Morrison
* fixed bug where small raytraces were being rotated - Erik Greenwald
* ios-icon.sh script added for generating IOS icons - Erik Greenwald
* support for implicit to boundary representation geometry conversion
	- Cliff Yapp, Wu Jianbang, Sean Morrison

src/rt/reshoot.c
/*                       R E S H O O T . C
 * BRL-CAD
 *
 * Copyright (c) 2004-2012 United States Government as represented by
 * the U.S. Army Research Laboratory.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * version 2.1 as published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this file; see the file named COPYING for more
 * information.
 */
/** @file rt/reshoot.c
 *
 * A program to shoot reshoot and compare results to previous runs.
 *
 * This program is geared towards performing regression tests of the
 * BRL-CAD libraries.  A candidate application is run with
 * the RT_G_DEBUG flag DEBUG_ALLHITS set.  This causes the application
 * to log all calls to the application hit() routine, and print the
 * contents of the struct partition.  The log is processed to
 * eliminate extraneous information to produce the input to this
 * program.  The following awk program, will suffice:  @verbatim

 /Pnt/ { START=index($0, "(") + 1
 STR=substr($0, START, index($0, ")") - START)
 gsub(  ", ", ",", STR)
 printf "Pnt=%s\n", STR

 }


 /Dir/ { START=index($0, "(") + 1
 STR=substr($0, START, index($0, ")") - START)
 gsub(  ", ", ",", STR)
 printf "Dir=%s\n", STR
 }
 /PT/  { PARTIN=$3
 PARTOUT=$5
 }
 /InHIT/ { INHIT=$2 }
 /OutHIT/ { OUTHIT=$2 }
 /Region/ { printf "\tregion=%s in=%s in%s out=%s out%s\n", $2, PARTIN, INHIT, PARTOUT, OUTHIT}

 @endverbatim
 * If this awk program is stored in the file p.awk then: @verbatim
 awk -f p.awk < logfile > inputfile
 @endverbatim
 * will produce a suitable input file for this program.  The form is as
 * follows: @verbatim
 Pnt=1, 2, 3
 Dir=4, 5, 6
 region=/all.g/platform.r in=platform.s indist=10016.8 out=platform.s outdist=10023.8
 @endverbatim
 * where the line beginning with "region" may be repeated any number of times, representing each
 * region encountered along the ray.
 * now run this program as follows: @verbatim
 reshoot geom.g obj [obj...] < inputfile
 @endverbatim
 * and this  will re-shoot all of the rays that the original program
 * shot, and compare the results.
 *
 * One of the basic assumptions is that the application structure
 * field a_onehit is set to zero in the original application, causing
 * all rays to be shot all the way through the geometry
 */

#include "common.h"

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <math.h>
#include <stddef.h>

#include "vmath.h"		/* vector math macros */
#include "bu.h"
#include "raytrace.h"		/* librt interface definitions */

#include "./rtuif.h"


char *progname = "(noname)";

/**
 * @struct shot
 * A description of a single shotline, and all the regions that were
 * hit by the shotline.
 */
struct shot {
    point_t pt;
    vect_t dir;
    struct bu_list regions;
};

/**
 * @struct shot_sp
 * The parse table for a struct shot
 */
static const struct bu_structparse shot_sp[] = {
    { "%f", 3, "Pnt", bu_offsetofarray(struct shot, pt), BU_STRUCTPARSE_FUNC_NULL, NULL, NULL},
    { "%f", 3, "Dir", bu_offsetofarray(struct shot, dir), BU_STRUCTPARSE_FUNC_NULL, NULL, NULL},
    {"", 0, (char *)0, 0, BU_STRUCTPARSE_FUNC_NULL, NULL, NULL }
};

/**
 * @struct reg_hit
 *
 * This describes the interaction of the ray with a single region
 */
struct reg_hit {
    struct bu_list l; /* linked list membership */
    struct bu_vls regname; /* name of the region */
    struct bu_vls in_primitive; /* name of the primitive for the inbound hit */
    double indist; /* distance along ray to the inbound hit */
    struct bu_vls out_primitive; /* name of the primitive for the outbound hit */
    double outdist; /* distance along ray to the outbound hit */
};

static const struct bu_structparse reg_sp[] = {
    {"%V", 1, "region", bu_offsetof(struct reg_hit, regname), BU_STRUCTPARSE_FUNC_NULL, NULL, NULL },
    {"%V", 1, "in", bu_offsetof(struct reg_hit, in_primitive), BU_STRUCTPARSE_FUNC_NULL, NULL, NULL},
    {"%V", 1, "out", bu_offsetof(struct reg_hit, out_primitive), BU_STRUCTPARSE_FUNC_NULL, NULL, NULL},
    {"%f", 1, "indist", bu_offsetof(struct reg_hit, indist), BU_STRUCTPARSE_FUNC_NULL, NULL, NULL},
    {"%f", 1, "outdist", bu_offsetof(struct reg_hit, outdist), BU_STRUCTPARSE_FUNC_NULL, NULL, NULL},
    {"", 0, (char *)0, 0, BU_STRUCTPARSE_FUNC_NULL, NULL, NULL }
};


/**
 *	U S A G E --- tell user how to invoke this program, then exit
 */
void
usage(char *s)
{
    if (s) (void)fputs(s, stderr);
    bu_exit(1, "Usage: %s geom.g obj [obj...] < rayfile \n", progname);
}


/**
 *  Process a single ray intersection list.
 *  A pointer to this function is in the application structure a_hit field.
 *  rt_shootray() calls this if geometry is hit by the ray.  It passes the
 *  application structure which describes the state of the world
 *  (see raytrace.h), and a circular linked list of partitions,
 *  each one describing one in and out segment of one region.
 * @return
 *	integer value, typically 1.  This value becomes the return value to rtshootray()
 */
int
hit(struct application *ap, struct partition *PartHeadp, struct seg *UNUSED(segs))
{
    /* see raytrace.h for all of these guys */
    register struct partition *pp;
    struct shot *sh = (struct shot *)ap->a_uptr;
    int status = 0;
    struct reg_hit *rh;
    struct bu_vls v = BU_VLS_INIT_ZERO;
    struct bu_vls result = BU_VLS_INIT_ZERO;
    struct valstruct {
	double val;
    } vs;
    static struct bu_structparse val_sp[] = {
	{"%f", 1, "val", bu_offsetof(struct valstruct, val), BU_STRUCTPARSE_FUNC_NULL, NULL, NULL},
    };

    /* examine each partition until we get back to the head */
    rh = BU_LIST_FIRST(reg_hit, &sh->regions);
    for ( pp=PartHeadp->pt_forw; pp != PartHeadp; pp = pp->pt_forw )  {

	bu_vls_trunc(&result, 0);

	/* since the values were printed out using a %g format,
	 * we have to do the same thing to the result to compare them
	 */
	bu_vls_trunc(&v, 0);
	bu_vls_printf(&v, "val=%g", pp->pt_inhit->hit_dist);
	if (bu_struct_parse(&v, val_sp, (const char *)&vs) < 0) {
	  bu_log("Warning - bu_struct_parse failure in reshoot.c, function hit\n");
	}

	if (!EQUAL(vs.val, rh->indist)) {
	    bu_vls_printf(&result, "\tinhit mismatch %g %g\n", pp->pt_inhit->hit_dist, rh->indist);
	    status = 1;
	}

	bu_vls_trunc(&v, 0);
	bu_vls_printf(&v, "val=%g", pp->pt_outhit->hit_dist);
	if (bu_struct_parse(&v, val_sp, (const char *)&vs) < 0) {
	  bu_log("Warning - bu_struct_parse failure in reshoot.c, function hit\n");
	}

	if (!EQUAL(vs.val, rh->outdist)) {
	    bu_vls_printf(&result, "\touthit mismatch %g %g\n", pp->pt_outhit->hit_dist, rh->outdist);
	    status = 1;
	}

	/* check the region name */
	if (!BU_STR_EQUAL( pp->pt_regionp->reg_name, bu_vls_addr(&rh->regname) )) {
	    /* region names don't match */
	    bu_vls_printf(&result, "\tregion name mismatch %s %s\n", pp->pt_regionp->reg_name, bu_vls_addr(&rh->regname) );
	    status = 1;
	}

	if ( !BU_STR_EQUAL(pp->pt_inseg->seg_stp->st_dp->d_namep, bu_vls_addr(&rh->in_primitive))) {
	    bu_vls_printf(&result, "\tin primitive name mismatch %s %s\n", pp->pt_inseg->seg_stp->st_dp->d_namep, bu_vls_addr(&rh->in_primitive));
	    status = 1;
	}
	if ( !BU_STR_EQUAL(pp->pt_outseg->seg_stp->st_dp->d_namep, bu_vls_addr(&rh->out_primitive))) {
	    bu_vls_printf(&result, "\tout primitive name mismatch %s %s\n", pp->pt_outseg->seg_stp->st_dp->d_namep, bu_vls_addr(&rh->out_primitive));
	    status = 1;
	}
	if (bu_vls_strlen(&result) > 0) {
	    bu_log("Ray Pt %g,%g,%g Dir %g,%g,%g\n%V",
		   V3ARGS(sh->pt),
		   V3ARGS(sh->dir),
		   &result);
	}

	rh = BU_LIST_NEXT(reg_hit, &rh->l);
    }

    /*
     * This value is returned by rt_shootray
     * a hit usually returns 1, miss 0.
     */
    return status;
}


/**
 * Function called when ray misses all geometry
 * A pointer to this function is stored in the application structure
 * field a_miss.  rt_shootray() will call this when the ray misses all geometry.
 * it passes the application structure.
 * @return
 *	Typically 0, and becomes the return value from rt_shootray()
 */
int
miss(register struct application *UNUSED(ap))
{
    return 0;
}

/**
 * Print a shot.  Mostly a debugging routine.
 *
 */
void
pr_shot(struct shot *sh)
{
    struct reg_hit *rh;
    /* shoot the old ray */
    bu_struct_print("shooting", shot_sp, (const char *)sh);
    for (BU_LIST_FOR(rh, reg_hit, &sh->regions)) {
	bu_struct_print("", reg_sp, (const char *)rh);
    }
}

/**
 *	Re-shoot a ray
 *	The ray described by the parametry sh
 *
 *	@param sh  a pointer to a struct shot describing the ray to shoot and expected results
 *	@param ap  a pointer to a struct application
 *	@return
 *	integer value indicating if there was a difference
 */
int
do_shot(struct shot *sh, struct application *ap)
{
    struct reg_hit *rh;
    int status;


    VMOVE(ap->a_ray.r_pt, sh->pt);
    VMOVE(ap->a_ray.r_dir, sh->dir);
    ap->a_uptr = (genptr_t)sh;

    ap->a_hit = hit;
    ap->a_miss = miss;
    status = rt_shootray( ap );	/* do it */

    /* clean up */
    while (BU_LIST_WHILE(rh, reg_hit, &sh->regions)) {
	BU_LIST_DEQUEUE( &(rh->l) );
	bu_vls_free(&rh->regname);
	bu_vls_free(&rh->in_primitive);
	bu_vls_free(&rh->out_primitive);
	bu_free(rh, "");
    }
    return status;
}


/**
 *	Load the database, parse the input, and shoot the rays
 * @return
 * integer flag value indicating whether any differences were detected.  0 = none, !0 = different
 */
int
main(int argc, char **argv)
{
    /* every application needs one of these */
    struct application	ap;
    static struct rt_i *rtip;	/* rt_dirbuild returns this */
    char idbuf[2048] = {0};	/* First ID record info */

    int status = 0;
    struct bu_vls buf = BU_VLS_INIT_ZERO;
    struct shot sh;

    progname = argv[0];

    if (argc < 3) {
	usage("insufficient args\n");
    }


    /*
     *  Load database.
     *  rt_dirbuild() returns an "instance" pointer which describes
     *  the database to be ray traced.  It also gives you back the
     *  title string in the header (ID) record.
     */
    if ( (rtip=rt_dirbuild(argv[1], idbuf, sizeof(idbuf))) == RTI_NULL ) {
	bu_exit(2, "rtexample: rt_dirbuild failure\n");
    }

    /* initialize the application structure to all zeros */
    RT_APPLICATION_INIT(&ap);

    ap.a_rt_i = rtip;	/* your application uses this instance */

    /* Walk trees.
     * Here you identify any object trees in the database that you
     * want included in the ray trace.
     */
    while (argc > 2) {
	if ( rt_gettree(rtip, argv[2]) < 0 )
	    fprintf(stderr, "rt_gettree(%s) FAILED\n", argv[0]);
	argc--;
	argv++;
    }
    /*
     * This next call gets the database ready for ray tracing.
     * (it precomputes some values, sets up space partitioning, etc.)
     */
    rt_prep_parallel(rtip, 1);

    memset((void *)&sh, 0, sizeof(sh));
    BU_LIST_INIT(&sh.regions);

    while (bu_vls_gets(&buf, stdin) >= 0) {
	char *p = bu_vls_addr(&buf);

	switch (*p) {
	    case 'P' :
	    {

		if (BU_LIST_NON_EMPTY(&sh.regions)) {
		    status |= do_shot(&sh, &ap);
		}

		if (bu_struct_parse(&buf, shot_sp, (const char *)&sh)) {
		    bu_exit(EXIT_FAILURE, "error parsing pt");
		}

		break;
	    }
	    case 'D' :
	    {
		if (bu_struct_parse(&buf, shot_sp, (const char *)&sh)) {
		    bu_exit(EXIT_FAILURE, "error parsing dir");
		}
		break;
	    }

	    default:
	    {
		struct reg_hit *rh = bu_calloc(1, sizeof (struct reg_hit), "");
		BU_VLS_INIT(&rh->regname);
		BU_VLS_INIT(&rh->in_primitive);
		BU_VLS_INIT(&rh->out_primitive);

		if (bu_struct_parse(&buf, reg_sp, (const char *)rh)) {
		    bu_log("Error parsing region %s\nSkipping to next line\n",
			   bu_vls_addr(&buf));
		}
		BU_LIST_APPEND(&sh.regions, &rh->l);

		break;
	    }
	}
	bu_vls_trunc(&buf, 0);
    }
    status |= do_shot(&sh, &ap);


    return status;
}


/*
 * Local Variables:
 * mode: C
 * tab-width: 8
 * indent-tabs-mode: t
 * c-file-style: "stroustrup"
 * End:
 * ex: shiftwidth=4 tabstop=8
 */
